### PR TITLE
GH-41478: [C++] Clean up more redundant move warnings

### PR DESCRIPTION
**Pull request by WillAyd**
### Rationale for this change

Minor warning cleanup for downstream libraries trying to get warning-free builds

### What changes are included in this PR?

Removed redundant std::move from return statements

### Are these changes tested?

Builds cleanly

### Are there any user-facing changes?

No

* GitHub Issue: #41478

### CODE REVIEW

**Comment by WillAyd** [comment]
Can see this in a branch adding meson warning level 2 downstream to nanoarrow:

https://github.com/apache/arrow-nanoarrow/actions/runs/8912807091/job/24476999593?pr=448

**Comment by kou** [comment]
@github-actions crossbow submit -g cpp

**Comment by kou** [comment]
Hmm. Why is `-Wredundant-move` warning detected in our CI?
It seems that nanoarrow uses `warning_level=2` https://github.com/apache/arrow-nanoarrow/blob/ae84d5f1f195f7345fe27d00e09507958746590e/meson.build#L27 . Is it related? If so, should we use `-Wextra` or `-Wredundant-move` in our CMake configuration?

**Comment by WillAyd** [comment]
I think you are asking why redundant move is _not_ is the Arrow CI but correct me if mistaken. Yea Meson's warning level 2 is essentially Wextra/Wall for gcc/clang. 

I haven't fully sifted through the current Arrow CMake configuration but I see that there is a BUILD_WARNING_LEVEL for CHECKIN that should pair both Wall and Wextra. I couldn't find that in the workflows - maybe it is not being used?

**Comment by kou** [comment]
You're right... I missed "not"... Sorry...

Could you try this whether we can detect these warnings in our build?

```diff
diff --git a/cpp/cmake_modules/SetupCxxFlags.cmake b/cpp/cmake_modules/SetupCxxFlags.cmake
index ea357b4779..a7001d6774 100644
--- a/cpp/cmake_modules/SetupCxxFlags.cmake
+++ b/cpp/cmake_modules/SetupCxxFlags.cmake
@@ -330,8 +330,9 @@ if("${BUILD_WARNING_LEVEL}" STREQUAL "CHECKIN")
     set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wall")
     set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wno-conversion")
     set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wno-sign-conversion")
-    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wunused-result")
     set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wdate-time")
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wredundant-move")
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wunused-result")
   elseif(CMAKE_CXX_COMPILER_ID STREQUAL "Intel" OR CMAKE_CXX_COMPILER_ID STREQUAL
                                                    "IntelLLVM")
     if(WIN32)
```

**Comment by WillAyd** [comment]
So I've added that but it flags a lot of issues in the current code base. What I have so far generates a clean build with the `ninja-debug` preset but there are still issues; guessing they come from full builds.

I'm happy to clean those up as well just wanted to confirm it isn't too big of a change in its current state with you before continuing.

Another possible issue is that the `-Wredundant-move` flag also gets applied to non-cpp files, yielding the following warnings during a build:

```sh
[140/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/musl/strptime.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[147/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriIp4Base.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[148/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriIp4.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[149/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriMemory.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[150/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriCompare.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[151/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriCommon.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[152/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriNormalizeBase.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[153/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriParseBase.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[154/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriEscape.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[155/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriNormalize.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[156/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriResolve.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[157/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriShorten.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[159/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriFile.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[160/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriRecompose.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[162/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriQuery.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
[164/780] Building C object src/arrow/CMakeFiles/arrow_vendored.dir/vendored/uriparser/UriParse.c.o
cc1: warning: command-line option ‘-Wredundant-move’ is valid for C++/ObjC++ but not for C
```

I didn't quite where those flags were getting intermixed yet; something to investigate

**Comment by kou** [comment]
Ah, sorry. Could you use `CXX_ONLY_FLAGS` instead?

```diff
diff --git a/cpp/cmake_modules/SetupCxxFlags.cmake b/cpp/cmake_modules/SetupCxxFlags.cmake
index ea357b4779..0bfb1e5a29 100644
--- a/cpp/cmake_modules/SetupCxxFlags.cmake
+++ b/cpp/cmake_modules/SetupCxxFlags.cmake
@@ -332,6 +332,7 @@ if("${BUILD_WARNING_LEVEL}" STREQUAL "CHECKIN")
     set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wno-sign-conversion")
     set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wunused-result")
     set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wdate-time")
+    string(APPEND CXX_ONLY_FLAGS " -Wredundant-move")
   elseif(CMAKE_CXX_COMPILER_ID STREQUAL "Intel" OR CMAKE_CXX_COMPILER_ID STREQUAL
                                                    "IntelLLVM")
     if(WIN32)
```

**Comment by WillAyd** [comment]
@github-actions crossbow submit -g cpp

**Comment by kou** [comment]
@github-actions crossbow submit -g cpp

**Comment by kou** [comment]
How about finding `cmake ...` line?
For example: https://github.com/ursacomputing/crossbow/actions/runs/9025575261/job/24801509542#step:7:244

Or how about build configuration summary at the end of `cmake ...` output?
For example: https://github.com/ursacomputing/crossbow/actions/runs/9025575261/job/24801509542#step:7:722

**Comment by WillAyd** [comment]
OK think those are fixed up. I believe gcc-13 was required to detect those

**Comment by kou** [comment]
@github-actions crossbow submit -g cpp

**Comment by WillAyd** [comment]
Don't have a CUDA device to locally test that. Fixed the latest crossbow issue - hopefully that covers it

**Comment by kou** [comment]
@github-actions crossbow submit -g cpp

**Comment by WillAyd** [comment]
I don't believe the CI failures are related - maybe flaky tests?

File: cpp/src/arrow/dataset/dataset_writer.cc [hunk 1]
```diff
@@ -409,7 +409,7 @@ class DatasetWriterDirectoryQueue {
     dir_queue->PrepareDirectory();
     ARROW_ASSIGN_OR_RAISE(dir_queue->current_filename_, dir_queue->GetNextFilename());
     // std::move required to make RTools 3.5 mingw compiler happy
```
Comment (kou) [inline cpp/src/arrow/dataset/dataset_writer.cc:411]: Could you remove this comment?

File: cpp/src/arrow/io/compressed.cc [hunk 2]
```diff
@@ -413,7 +413,7 @@ class CompressedInputStream::Impl {
     RETURN_NOT_OK(buf->Resize(bytes_read));
     // Using std::move because some compiler might has issue below:
     // https://wg21.cmeerw.net/cwg/issue1579
```
Comment (kou) [inline cpp/src/arrow/io/compressed.cc:415]: Hmm. We may not be able to remove this `std::move()`: https://github.com/apache/arrow/pull/39807

Let's try R jobs.

File: cpp/src/arrow/util/iterator.h [hunk 3]
```diff
@@ -180,9 +180,8 @@ class Iterator : public util::EqualityComparable<Iterator<T>> {
       ARROW_ASSIGN_OR_RAISE(auto element, maybe_element);
       out.push_back(std::move(element));
     }
-    // ARROW-8193: On gcc-4.8 without the explicit move it tries to use the
-    // copy constructor, which may be deleted on the elements of type T
-    return std::move(out);
+
```
Comment (kou) [inline cpp/src/arrow/util/iterator.h:183]:
```suggestion
```

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
`test-r-rstudio-r-base-4.1-opensuse153` failed:

https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=63897&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c939d89-0d1a-51f2-8b30-091a7a82e98c&l=1320

```text
[ 26%] Building CXX object src/arrow/CMakeFiles/arrow_util.dir/util/align_util.cc.o
/arrow/cpp/src/arrow/util/align_util.cc: In function ‘arrow::Result<std::shared_ptr<arrow::Buffer> > arrow::util::EnsureAlignment(std::shared_ptr<arrow::Buffer>, int64_t, arrow::MemoryPool*)’:
/arrow/cpp/src/arrow/util/align_util.cc:162:12: error: could not convert ‘new_buffer’ from ‘std::unique_ptr<arrow::Buffer>’ to ‘arrow::Result<std::shared_ptr<arrow::Buffer> >’
     return new_buffer;
            ^~~~~~~~~~
make[2]: *** [src/arrow/CMakeFiles/arrow_util.dir/build.make:80: src/arrow/CMakeFiles/arrow_util.dir/util/align_util.cc.o] Error 1
```

Let's update `test-r-rstudio-r-base-4.1-opensuse153`: https://github.com/apache/arrow/issues/41626

**Comment by kou** [comment]
Could you rebase on main to use OpenSUSE 15.5 not 15.3?

**Review by westonpace — APPROVED** [review]
+1 (providing CI / nightly tests pass)  These look like nice improvements to me.  I think the main reason we had the move on return in the first place was due to a compiler bug in the old R 3.5 toolchain.  Now that we no longer need to build there I think this is a good cleanup.

**Comment by kou** [comment]
@github-actions crossbow submit test-r-rstudio-r-base-4.1-opensuse155

**Comment by kou** [comment]
@github-actions crossbow submit -g cpp -g r

**Comment by WillAyd** [comment]
Well this is fun...looks like the compilers have conflicting desires. With code like this:

```cpp
template <typename Options, typename... Properties>
const FunctionOptionsType* GetFunctionOptionsType(const Properties&... properties) {
  ...
  auto options = std::make_unique<Options>();
  RETURN_NOT_OK(
      FromStructScalarImpl<Options>(options.get(), scalar, properties_).status_);

   return std::move(options);
}
```

The more modern compilers complain about a redundant move:

```sh
/home/willayd/clones/arrow/cpp/src/arrow/compute/function_internal.h: In instantiation of ‘arrow::Result<std::unique_ptr<arrow::compute::FunctionOptions> > arrow::compute::internal::GetFunctionOptionsType(const Properties& ...)::OptionsType::FromStructScalar(const arrow::StructScalar&) const [with Options = arrow::compute::ArithmeticOptions; Properties = {arrow::internal::DataMemberProperty<arrow::compute::ArithmeticOptions, bool>}]’:
/home/willayd/clones/arrow/cpp/src/arrow/compute/function_internal.h:697:3:   required from ‘const arrow::compute::FunctionOptionsType* arrow::compute::internal::GetFunctionOptionsType(const Properties& ...) [with Options = arrow::compute::ArithmeticOptions; Properties = {arrow::internal::DataMemberProperty<arrow::compute::ArithmeticOptions, bool>}]’
/home/willayd/clones/arrow/cpp/src/arrow/compute/api_scalar.cc:314:79:   required from here
/home/willayd/clones/arrow/cpp/src/arrow/compute/function_internal.h:687:31: error: redundant move in return statement [-Werror=redundant-move]
  687 |       return std::move(options);
```

If you remove the `std::move(...)` then the openSUSE builds complain that they cannot convert the templated pointer type to that which matches the function declaration:

```sh
  /arrow/cpp/src/arrow/compute/function_internal.h:687:14: error: could not convert ‘options’ from ‘std::unique_ptr<arrow::compute::ArithmeticOptions, std::default_delete<arrow::compute::ArithmeticOptions> >’ to ‘arrow::Result<std::unique_ptr<arrow::compute::FunctionOptions> >’
       return options;
```

This also seems to happen in cases where the returned value is a unique_ptr but the function/method declaration is that of a shared_ptr; I'm assuming more modern compilers happily convert that automatically for you but the compiler on the openSUSE build does not

For now I've tried to be really explicit about the return type in the offending functions to try and appease all platforms, but happy to change to any better idea

**Comment by kou** [comment]
@github-actions crossbow submit -g cpp -g r

**Comment by WillAyd** [comment]
Looks like the valgrind failures are from AWS cryptography code - I don't think that is related to this

**Review by kou — APPROVED** [review]
+1

@pitrou @bkietz Do you want to review this before we merge this?

**Review by pitrou — APPROVED** [review]
+1, thanks for doing this

**Review by bkietz — APPROVED** [review]
Thanks for doing this!

**Pull request merged**